Tflite GPU OpenGL Delegate Allow HWC Const Add

svobora opened the pull request:
Allow addition (Add) of constant HWC tensor in OpenGL GPU delegate of Tensorflow Lite.

svobora committed:
Tflite GPU OpenGL Delegate Allow HWC Const Add

Allow addition of constant HWC tensor in OpenGL GPU delegate of Tensorflow Lite.

impjdi commented on tensorflow/lite/delegates/gpu/gl/kernels/add.cc:
```diff
@@ -102,6 +99,28 @@ class Add : public NodeShader {
           /*input=*/IOStructure::AUTO,
           /*output=*/IOStructure::AUTO,
       };
```
return early:

return absl::OkStatus();

and when you do that, you don't need to keep a long if else if else sequence.  just break off the else if, and do:

if (hwc_tensor) {

also, do an early return there as well.  you can remove the final else block.

svobora replied:
The original code does not return early, I did not modify it.

impjdi commented on tensorflow/lite/delegates/gpu/gl/kernels/add.cc:
```diff
@@ -102,6 +99,28 @@ class Add : public NodeShader {
+        /*source_code=*/"value_0 += $hwc_buffer[gid.x, gid.y, gid.z]$;",
+        /*input=*/IOStructure::AUTO,
+        /*output=*/IOStructure::AUTO,
+        };
```
indentation is off.

svobora replied:
True.

impjdi commented on tensorflow/lite/delegates/gpu/gl/kernels/add.cc:
```diff
@@ -102,6 +99,28 @@ class Add : public NodeShader {
       };
+    } else if (hwc_tensor) {
+      *generated_code = {
+        /*parameters=*/{},
```
indentation in the {} block needs to be 4, but you indented 2.  so 2 more spaces from L.104 to L.122.

svobora replied:
The whole file has indentation of 2. So I kept it, if you want to fix indentation use a script.

impjdi replied:
Sorry, we don't follow up with "let the external user submit whatever code they want, and i'll clean it up later with a beautifier."  I suggest you clean it up yourself with `clang-format` with `-style Google`.

svobora commented:
@impjdi If you want to refactor code of add.cc to your new coding conventions, like e.g. in mul.cc, this is an independent issue. Until that is done, I do not feel like it to have a single code file with multiple coding conventions.

impjdi commented:
While I agree in general that your change should only modify the least amount of change necessary, it's also true that your code adds decent amount of lines and make it harder to read with 3 variables instead of 2.  That should probably go together.

svobora committed:
Refactor to tf+google guidelines

4-space blocks, 80 chars max per line, return ASAP
commit readability decreased significantly

svobora commented:
Code is refactored as requested. The readability of the commit is not 0.

impjdi commented on tensorflow/lite/delegates/gpu/gl/kernels/add.cc:
```diff
@@ -35,116 +35,128 @@ namespace {
+                {{"hwc_buffer",
+                MakeReadonlyObject(
+                    uint3(static_cast<int>(ctx.input_shapes[0][2]),
+                        static_cast<int>(ctx.input_shapes[0][1]),
```
oh, no.  you misunderstood my comment.  I didn't ask the entire file to be indented by 4, but to indent 

*generated_code = {
// indent by 4 here instead of just 2
};

The rest of the code should be just indent by 2.  Usually, clang-format -style Google will be able to deal with all this.

svobora replied:
ups

svobora committed:
Revert, set default indent 2, wrapped param indent to 4

svobora commented:
lets see if third time's the charm

impjdi commented on tensorflow/lite/delegates/gpu/gl/kernels/add.cc:
```diff
@@ -1,11 +1,8 @@
 /* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
-
```
you accidentally removed the blank lines in L.2, 6, and 8.  please revert.

impjdi reviewed (changes requested):
Looks good.  Only revert the removed blank lines in the copyright comments, and you're good to go!

svobora committed:
Revert blank lines in copyright

svobora committed:
Update BUILD add.cc dep

svobora commented:
Bazel GPU CI build failed probably due to missing new dep, hopefully fixed. (I build using cmake)

svobora commented:
Has there been some issue with copybara? Looks like all pull requests were affected but now it should be fixed.

impjdi commented:
I received your PR in the internal script which I have to approve as well, but it only includes the `add.cc` change without the change in the `BUILD` file.  I'm not familiar with the github UI, but can you make sure that both files are exported?

svobora commented:
Any idea on what exactly am I supposed to do? I suppose it is the copybara's task to copy the changed files into the internal system. Maybe rerun the copybara task? Other approach would be to make new PR with the final changes in a single commit.

impjdi commented:
@gbaned Do you know what needs to be done?

svobora commented:
Possibly the same issue happened at the same time with [#PR51866](https://github.com/tensorflow/tensorflow/pull/51866), fixed by @cheshire (by rerunning kokoro:force-run?).
